pkg/reconciler/brokercell/brokercell_test.go

```diff
@@ -933,3 +906,64 @@ func emptyHPASpec(template *hpav2beta2.HorizontalPodAutoscaler) *hpav2beta2.Hori
+// and trigger. Since the serialization order of the binary data of brokerTargets in the configMap is not guaranteed, we need
+// to deserialization the binary data to a brokerTargets proto to compare, so it should be rewritten without using the tableTest Utility.
+func TestBrokerTargetsReconcileConfig(t *testing.T) {
+	setReconcilerEnv()
+	bc := NewBrokerCell(brokerCellName, testNS, WithBrokerCellSetDefaults)
+	objects:= []runtime.Object{
+		bc,
+		NewBroker("broker", testNS, WithBrokerSetDefaults),
+		NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
+		NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults),
+	}
+	ctx, _ := SetupFakeContext(t)
+	cmw := configmap.NewStaticWatcher()
+	ctx, client := fakekubeclient.With(ctx,)
+	base := reconciler.NewBase(ctx, controllerAgentName, cmw)
+	testingListers := NewListers(objects)
+	ls := listers{
+		brokerLister:     testingListers.GetBrokerLister(),
+		hpaLister:        testingListers.GetHPALister(),
+		triggerLister:    testingListers.GetTriggerLister(),
+		configMapLister:  testingListers.GetConfigMapLister(),
+		serviceLister:    testingListers.GetK8sServiceLister(),
+		endpointsLister:  testingListers.GetEndpointsLister(),
+		deploymentLister: testingListers.GetDeploymentLister(),
+		podLister:        testingListers.GetPodLister(),
+	}
+	r, err := NewReconciler(base, ls)
+	if err != nil {
+		t.Fatalf("Failed to created BrokerCell reconciler: %v", err)
+	}
+	// here we only want to test the functionality of the reconcileConfig that it should create a brokerTargets config successfully
+	r.reconcileConfig(ctx, bc)
+	wantMap := testingdata.Config(t, NewBrokerCell(brokerCellName, testNS, WithBrokerCellSetDefaults),
+		NewBroker("broker", testNS, WithBrokerSetDefaults),
+		NewTrigger("trigger1", testNS, "broker", WithTriggerSetDefaults),
+		NewTrigger("trigger2", testNS, "broker", WithTriggerSetDefaults))
+	mapList, err := client.CoreV1().ConfigMaps(testNS).List(metav1.ListOptions{})
+	if err != nil || len(mapList.Items) != 1 {
+		t.Fatalf("Failed to get configMap: %v", err)
+	}
+	gotMap := mapList.Items[0]
+	// compare the ObjectMeta field
+	if diff := cmp.Diff(wantMap.ObjectMeta, gotMap.ObjectMeta, cmpopts.EquateEmpty()); diff != "" {
+		t.Fatalf("Unexpected ObjectMeta in ConfigMap(-want, +got): %s", diff)
+	}
+	// deserialize the binary data to a broker targets config proto
+	var wantBrokerTargets config.TargetsConfig
+	var gotBrokerTargets config.TargetsConfig
+	if err := proto.Unmarshal(wantMap.BinaryData[targetsCMKey], &wantBrokerTargets); err != nil {
+		t.Fatalf("Failed to deserialize the binary data in configMap: %v", err)
+	}
+	if err := proto.Unmarshal(gotMap.BinaryData[targetsCMKey], &gotBrokerTargets); err != nil {
+		t.Fatalf("Failed to deserialize the binary data in configMap: %v", err)
+	}
+	// compare the broker targets config
+	if diff := cmp.Diff(wantBrokerTargets.String(), gotBrokerTargets.String(), cmpopts.EquateEmpty()); diff != "" {
```

Comment:
Have you verified that `String()` maintains the order of the map? I would expect cmp.Diff(wantBrokerTargets, gotBrokerTargets) should do the trick.

Reply:
Yeah, the `String()` maintains the order, as I rerun the fixed test 100 times and they all passed. I tried to compare the targetsConfig type directly but got this error, ```panic: cannot handle unexported field at {config.TargetsConfig}.state:
	"github.com/google/knative-gcp/pkg/broker/config".TargetsConfig
consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported [recovered]
```. It even happens after I added AllowUnexported and cmpopts.IgnoreUnexported as comparison option.

Reply:
I tried `proto.Equals()` and it works. The concerns is that this function only returns the boolean type, so I guess cmp.Diff() would be better when the two proto not matches as it will show the detailed diff?